File: src/main/java/dev/nytweetdeck/timeline/PostTextUrlNormalizer.java
package dev.nytweetdeck.timeline;

import java.net.URI;
import java.util.LinkedHashMap;
import java.util.List;
import dev.nytweetdeck.timeline.TimelinePage.TextLink;

final class PostTextUrlNormalizer {

    enum Kind {
        LINK,
        MEDIA,
        ARTICLE
    }

    record UrlEntity(
            String shortUrl,
            String displayUrl,
            String expandedUrl,
            String unwoundUrl,
            Kind kind) {

        UrlEntity(String shortUrl, String expandedUrl, String unwoundUrl, Kind kind) {
            this(shortUrl, null, expandedUrl, unwoundUrl, kind);
        }
    }

    private PostTextUrlNormalizer() {}

    static String normalize(String value, List<UrlEntity> entities) {
        if (value == null || entities.isEmpty()) {
            return value;
        }
        var replacements = new LinkedHashMap<String, String>();
        for (var entity : entities) {
            if (!isTcoUrl(entity.shortUrl())) {
                continue;
            }
            if (entity.kind() != Kind.LINK) {
                replacements.put(entity.shortUrl(), "");
                continue;
            }
            var destination = firstHttpUrl(entity.unwoundUrl(), entity.expandedUrl());
            if (destination != null && !isTcoUrl(destination)) {
                replacements.putIfAbsent(entity.shortUrl(), destination);
            }
        }
        var normalized = value;
        var removedContent = false;
        for (var replacement : replacements.entrySet()) {
            normalized = normalized.replace(replacement.getKey(), replacement.getValue());
            removedContent |= replacement.getValue().isEmpty();
        }
        if (!removedContent) {
            return normalized;
        }
        return normalized
                .replaceAll("[ \\t]{2,}", " ")
                .replaceAll("[ \\t]+(?=\\R|$)", "")
                .replaceAll("(?m)^[ \\t]+", "")
                .replaceAll("\\R{3,}", "\n\n")
                .strip();
    }

    static List<TextLink> links(List<UrlEntity> entities) {
        var links = new LinkedHashMap<String, TextLink>();
        for (var entity : entities) {
            if (entity.kind() != Kind.LINK) {
                continue;
            }
            var destination = firstHttpUrl(
                    entity.unwoundUrl(), entity.expandedUrl(), entity.shortUrl());
            if (destination == null) {
                continue;
            }
            var display = entity.displayUrl();
            if (display == null || display.isBlank()) {
                display = destination;
            }
            links.putIfAbsent(destination + "\0" + display, new TextLink(destination, display));
        }
        return List.copyOf(links.values());
    }

    private static String firstHttpUrl(String... candidates) {
        for (var candidate : candidates) {
            if (isHttpUrl(candidate)) {
                return candidate;
            }
        }
        return null;
    }

    private static boolean isTcoUrl(String value) {
        if (!isHttpUrl(value)) {
            return false;
        }
        return "t.co".equalsIgnoreCase(URI.create(value).getHost());
    }

    private static boolean isHttpUrl(String value) {
        if (value == null || value.isBlank()) {
            return false;
        }
        try {
            var uri = URI.create(value);
            return ("http".equalsIgnoreCase(uri.getScheme())
                            || "https".equalsIgnoreCase(uri.getScheme()))
                    && uri.getHost() != null;
        } catch (IllegalArgumentException ignored) {
            return false;
        }
    }
}

File: android/app/src/main/java/dev/nytweetdeck/android/xapi/PostTextUrlNormalizer.kt
package dev.nytweetdeck.android.xapi

import dev.nytweetdeck.android.model.TextLink
import java.net.URI

internal enum class PostUrlEntityKind {
    LINK,
    MEDIA,
    ARTICLE,
}

internal data class PostUrlEntity(
    val shortUrl: String,
    val expandedUrl: String?,
    val unwoundUrl: String?,
    val kind: PostUrlEntityKind,
    val displayUrl: String? = null,
)

internal object PostTextUrlNormalizer {
    fun normalize(value: String, entities: List<PostUrlEntity>): String {
        if (entities.isEmpty()) return value
        val replacements = linkedMapOf<String, String>()
        entities.forEach { entity ->
            if (!entity.shortUrl.isTcoUrl()) return@forEach
            if (entity.kind != PostUrlEntityKind.LINK) {
                replacements[entity.shortUrl] = ""
                return@forEach
            }
            val destination = sequenceOf(entity.unwoundUrl, entity.expandedUrl)
                .filterNotNull()
                .firstOrNull { it.isHttpUrl() }
            if (destination != null && !destination.isTcoUrl()) {
                replacements.putIfAbsent(entity.shortUrl, destination)
            }
        }
        var normalized = value
        var removedContent = false
        replacements.forEach { (shortUrl, replacement) ->
            normalized = normalized.replace(shortUrl, replacement)
            removedContent = removedContent || replacement.isEmpty()
        }
        return if (removedContent) {
            normalized
                .replace(EXCESSIVE_HORIZONTAL_SPACE, " ")
                .replace(TRAILING_HORIZONTAL_SPACE, "")
                .replace(LEADING_HORIZONTAL_SPACE, "")
                .replace(EXCESSIVE_LINE_BREAKS, "\n\n")
                .trim()
        } else {
            normalized
        }
    }

    fun links(entities: List<PostUrlEntity>): List<TextLink> = entities
        .asSequence()
        .filter { it.kind == PostUrlEntityKind.LINK }
        .mapNotNull { entity ->
            val destination = sequenceOf(entity.unwoundUrl, entity.expandedUrl, entity.shortUrl)
                .filterNotNull()
                .firstOrNull { it.isHttpUrl() }
                ?: return@mapNotNull null
            TextLink(destination, entity.displayUrl?.takeIf(String::isNotBlank) ?: destination)
        }
        .distinct()
        .toList()

    private fun String.isTcoUrl(): Boolean = isHttpUrl() &&
        runCatching { URI(this).host.equals("t.co", ignoreCase = true) }.getOrDefault(false)

    private fun String.isHttpUrl(): Boolean = runCatching {
        val uri = URI(this)
        (uri.scheme.equals("http", ignoreCase = true) || uri.scheme.equals("https", ignoreCase = true)) &&
            uri.host != null
    }.getOrDefault(false)

    private val EXCESSIVE_HORIZONTAL_SPACE = Regex("[ \\t]{2,}")
    private val TRAILING_HORIZONTAL_SPACE = Regex("[ \\t]+(?=\\R|$)")
    private val LEADING_HORIZONTAL_SPACE = Regex("(?m)^[ \\t]+")
    private val EXCESSIVE_LINE_BREAKS = Regex("\\R{3,}")
}
